File: 260-single-number-iii/single-number-iii.java
class Solution {
    public int[] singleNumber(int[] nums) {
        int finalXor = 0;

        for (int num : nums) {
            finalXor ^= num;
        }

        int group0 = 0;
        int group1 = 0;

        int bitPosDif = 0;

        while (((finalXor >> bitPosDif) & 1) != 1) {
            bitPosDif++;
        }

        for (int num: nums) {
            if (((num >> bitPosDif) & 1) == 1) {
                group1 ^= num;
            } else {
                group0 ^= num;
            }
        }
        return new int[]{group0, group1};
        
    }
}

File: 260-single-number-iii/single-number-iii.py
class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        final_xor: int = 0
        for x in nums:
            final_xor ^= x

        group0 = 0
        group1 = 0
        bit_pos_dif = 0
        while (final_xor >> bit_pos_dif) & 1 != 1:
            bit_pos_dif += 1
        
        for x in nums:
            if (x >> bit_pos_dif) & 1 == 1:
                group1 ^= x
            else:
                group0 ^= x
        return [group0, group1]



